Scrum 15 explore page be

alcarrillomunoz opened the pull request:
**Description**
This merge builds on previous itemRoutes CRUD functions developed by Tea and creates foundation for bartering process in backend. 

**Changes Made**
- authentication middleware was added to itemRoutes and barter routes for token authorization
- created controller for itemRoutes
- continued working on itemRoutes CRUD functions, and updated error handling according to team's input
- updates made to Item schema to create categories (item, service, lesson - per team meeting), and mark items as available (true/false when in barter process), and closed (true/false once traded)
- Barter data model was added
- barter route/controller was added for handling bartering backend

**Related Issues**
GET request for item filtering/searching is being worked on by Tea so I expect that may have some updates. The only thing I changed for that function was error handling (setting up api response per team input). We'll need to add filter edits to return only available = true items. 

**Additional Notes**
Testing was done via Postman.

**Merge Request Checklists**
Code follows project coding guidelines.
Documentation reflects the changes made.

orangesoftware commented on src/middleware/authentication.js:
Could you extract from the token the user Id and validate if user exists in the db, you have the User model but looks like is not used. What happen if token is expired?

alcarrillomunoz replied:
Hi Roberto, thanks for the feedback! I'm working on this and will follow up tomorrow.

orangesoftware commented on src/controllers/itemController.js:
Thank you for following the pattern response.

Miminiverse commented on src/models/Barter.js:
I have some questions and suggestions below: 
1. Instead of user1 and user2, would you consider using "initiator" and "recepient", that way you can also keep track of who start first. User1 and User2 seem to be confusing and unclear, especially in the frontend. Same for item 1 and item 2, maybe we can change to offeredItem and requestedItem. Names are important and clear, concise names will be helpful in the future.
2. Add required true for both users an items, I believe they are all should be required for a valid barter.

Miminiverse commented on src/models/item.js:
Both name and title seem redundant - any reason we need both?
Instead of having the boolean available and closed, would you consider having a status with enum (available, pending, closed), similar to what you have in the Barter Schema. That will be easier to keep track of the item. 
I believe assignedTo here is for the user who creates the item. If that's true, I'd suggest using "owner" instead of assisgnTo.

Miminiverse commented on src/models/User.js:
What are the tags for in User Schema? 
Would you consider adding tags in Item Schema? That will be useful.

Miminiverse commented on src/app.js:
You have /api/v1 and /api/items and api/barter - this mixes versioned and unversioned routes
Keep API routes under /api/v1/ prefix.
/api/v1/items
/api/v1/barter

alcarrillomunoz commented:
Thank you for the feedback Tam! The following changes have been made in latest commit: changed barter values to more descriptive titles (initiator, itemOffered, recipient, itemRequested), added required: true for both users and items in barter schema, changed itemStatus to be enum value with options: available, pending, closed, changed assignedTo value in item schema to owner, api/v1 routes updated to improve order

About the other notes: I'm not sure why name and title is used. It was set up this way when I pulled changes so maybe we can talk as a team tomorrow since there may be a reason I'm not aware of. And about tags, I believe that is something Tea and Elle are working on integrating with the search feature. We can talk about this as a team tomorrow as well to get more clarity if that sounds good?

Miminiverse commented on src/controllers/barter.js:
This check doesn't look logical to me, please check again. 
 if (
      (initiatorSearch._id.toString() !==
        offeredItemSearch.owner.toString() ||
        recipientSearch._id.toString() !==
          requestedItemSearch.owner.toString()) &&
      req.user.role !== "admin"
    )

Miminiverse replied:
this is okay, but it missed checking if the user is authorized to update or not, in this case anyone can update this barter.
    const barter = await Barter.findByIdAndUpdate(
      { _id: id },
      { status: status },
      { new: true, runValidators: true }
    );

Miminiverse replied:
Barter creation updates multiple documents (Barter, items). You can consider using Mongoose Transactions to avoid inconsistent states if one operation fails.

    await newBarter.save();
    await offeredItemSearch.save();
    await requestedItemSearch.save();

Miminiverse commented on src/controllers/itemController.js:
IN updateItem, we should check if the user is authorized to update the item or not first before we update the item. This is now coming after you update the item. 
 if (
      item.owner.toString() !== req.user.userId &&
      req.user.role !== "admin"
    ) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json(
          createResponse("error", "User is not authorized to edit this item")
        );
    }

alcarrillomunoz commented:
Changes made: only authorized users can update items and barter now, logic was corrected as well for user authorization checks in creating a new barter and update barter

I will look at adding Mongoose Transactions next, and adding JWT handling per reviewer feedback.

Miminiverse reviewed (approved):
Very nice job! Can't wait to see these new faetures in the app.
